cupy/_core/dlpack.pyx

```diff
@@ -19,139 +14,213 @@ import warnings
+    """
+    kind = dtype.kind
+
+    if kind == 'u':
+        return <uint8_t>kDLUInt
+    elif kind == 'i':
+        return <uint8_t>kDLInt
+    elif kind == 'f':
+        return <uint8_t>kDLFloat
+    elif kind == 'c':
+        return <uint8_t>kDLComplex
+    elif kind == 'b':
+        return <uint8_t>kDLBool
+    else:
+        raise BufferError('dtype is not supported for dlpack export')
+
 
-    cdef DLDevice* device = &dl_tensor.device
+cdef DLDevice get_dlpack_device(_ndarray_base array):
+    cdef DLDevice device
     cdef bint is_managed
-    cdef int dev_id = array.data.device_id
+
+    device.device_id = array.data.device_id
+
     if not runtime._is_hip_environment:
-        attrs = runtime.pointerGetAttributes(data_ptr)
+        attrs = runtime.pointerGetAttributes(array.data.ptr)
         is_managed = (attrs.type == runtime.memoryTypeManaged)
         if is_managed:
             device.device_type = kDLCUDAManaged
-            dev_id = 0  # make it accessible on CPU too
         else:
             device.device_type = kDLCUDA
     else:
         device.device_type = kDLROCM
-    device.device_id = dev_id
-
-    cdef DLDataType* dtype = &dl_tensor.dtype
-    if array.dtype.kind == 'u':
-        dtype.code = <uint8_t>kDLUInt
-    elif array.dtype.kind == 'i':
-        dtype.code = <uint8_t>kDLInt
-    elif array.dtype.kind == 'f':
-        dtype.code = <uint8_t>kDLFloat
-    elif array.dtype.kind == 'c':
-        dtype.code = <uint8_t>kDLComplex
-    elif array.dtype.kind == 'b':
-        dtype.code = <uint8_t>kDLBool
+
+    return device
+
+
+# The name of this function is following the framework integration guide of
+# TensorComprehensions.
+cpdef object toDlpack(
+    _ndarray_base array, bint use_versioned=True, bint to_cpu=False,
+    bint ensure_copy=False
+) except +:
+    cdef DLManagedTensor* dlm_tensor
+    cdef DLManagedTensorVersioned* dlm_tensor_ver
+    cdef DLTensor* dl_tensor
+    cdef const char *capsule_name
+
+    # Fetch dtype early (as this can raise a BufferError in theory)
+    cdef uint8_t dtype_code = get_dlpack_dtype_code(array.dtype)
+    cdef size_t dtype_itemsize = array.dtype.itemsize
+
+    # Fetch device information since we need it to deal with CPU logic.
+    cdef DLDevice device = get_dlpack_device(array)
+
+    cdef int32_t ndim = array._shape.size()
+
+    if not to_cpu:
+        owner = array
+    elif not ensure_copy and device.device_type == kDLCUDAManaged:
+        # Managed memory is CPU accessible but consumer may expect kDLCPU.
+        owner = array
+        device.device_type = kDLCPU
+        device.device_id = 0
     else:
-        raise ValueError('Unknown dtype')
-    dtype.lanes = <uint16_t>1
-    dtype.bits = <uint8_t>(array.dtype.itemsize * 8)
+        # We need to create a CPU copy.  Assumes owner.dtype == array.dtype.
+        # TODO: As noted at calling site, this does not honor the "stream".
+        owner = array.get(stream=None, order='A')
```

Comment:
This will be causing race conditions if host access follows right after hand-shaking. Need to sync internally:
```suggestion
        owner = array.get(order='A')  # this copies on the current stream and then sync
```

Reply:
Ah, thanks for the note.  IIRC, I added `stream=None`, because Cython complained about it.  Maybe it needs to be `stream=arr.stream`?
(but maybe I should just retry)

Reply:
Ahhh ok, your code was right, just need to remove the comment. You hit a Cython limitation when it comes to invoking cdef'd functions with non-default kw args.
https://github.com/cupy/cupy/blob/7cbf5f57b79ae047ec7774bf0568ca5fb3c561cb/cupy/_core/core.pyx#L1798

Reply:
Passing stream=None would mean to use the current stream.

Reply:
I now change my mind again and agree that this is a TODO. This is very tricky as I keep revisiting this part, and I'd like to share my understanding here to aid code review.

When this `toDlpack()` function was called, the producer/consumer streams are ordered, that's why we did not need to pass any stream argument. 

This is no longer true when CPU copy is requested. In the case of NumPy, [`stream=None` is provided](https://github.com/numpy/numpy/blob/816fd7c0d1569bef00b3765d47528b093cf55839/numpy/_core/src/multiarray/dlpack.c#L536), but CuPy turns it to `streamLegacy` in `__dlpack__` and orders it after CuPy's current stream (whatever it is), so we're now on the consumer's (legacy) stream. We need a way to pass this info to `get(stream=consumer_stream, order='A')` to make a blocking copy of the array on the legacy stream to CPU. The easiest way to ensure correctness is to pass it through `toDLPack()`.

Reply:
So, now that I think about it.  Let me recap two things:
1. The existing stream logic ensures that the GPU copy is available with respect to the consumer stream order.
2. If we just pass on `stream=consumer_stream` here, the copy to the CPU will also be correctly ordered w.r.t. to the consumer stream.

Lastly: If the user did not provide a `stream=`, we have to use `blocking=True` (default) to ensure the CPU copy is immediately accessible.  But we can pass `blocking=False` otherwise.

One other thing, though:  If we have managed memory, we actually need to add an explicit synchronization to ensure that the memory is CPU available?

Reply:
CPU users cannot sync unless they know where the data comes from, which means passing in the `stream` (which we support, even if it may be a bit fuzzy).

So, I think we must sync if a stream wasn't passed in, that is logic I need to add.  If a stream was passed in, we should copy on that stream (and that is all the sync we need).

Reply:
I have added some logic for this see last commit.  But it isn't really tested so needs careful review maybe.
It also is just generally confusing :(.  The first commit should be more straightforward addressing the review comments.

I am not familiar with how one may test the stream logic (the code paths are covered mostly at least, but nothing checks the stream order is right).